snapshot_iterator.go

```diff
@@ -163,6 +188,24 @@ func (s *snapshotIterator) Teardown(ctx context.Context) error {
 		return fmt.Errorf("failed to wait for snapshot acks: %w", err)
 	}
 
+	// waiting for the workers to finish will allow us to have an easier time
+	// debugging goroutine leaks.
+	_ = s.t.Wait()
+
+	sdk.Logger(ctx).Info().Msg("all workers done")
+
+	if _, err := s.config.db.ExecContext(ctx, "UNLOCK TABLES"); err != nil {
```

Comment:
You can unlock the tables earlier than this, if the snapshot is taking a long time, the tables will be unwritable for the duration. The lock needs to be in place for the duration when you get the master position and snapshot start/end values, after that you should be free to unlock the tables. 

I should add, keeping the lock for the whole duration of the snapshot, ensures the snapshot is always consistent.
I'm saying this, because deletes can sneak in and will never be recorded, though I think we are ok with this because it is a balance between performance and consistency.

Reply:
Oh so we are ok with having deletes / updates sneak in? those will be later captured from cdc mode, and that could be problematic couldn't it be?

I thought we wanted to prevent that at all costs.

Then again, if during the snapshot, a:

- delete happens: the destination that receives that delete record from the cdc iterator will try to delete data that doesn't exist. It isn't an error with `database/sql`, so I guess we should be fine.
- update happens: the destination would upsert data with a sdk.OperationUpdate, so there should also be no error.
- insert happens: the destination would upsert data with a sdk.OperationCreate, so there should also be no error.

So, now that I think about it, as you say, if we lock the tables only to get the master position we should be fine. @lyuboxa you say that it is a balance between performance and consistency.
Where is the consistency lost? I feel like at worst we are doing some redundant updates / deletes / inserts. If the whole table is updated while snapshotting, we don't really loose information (as long as the cdc iterator finishes streaming all those updates).

Reply:
> Oh so we are ok with having deletes / updates sneak in? those will be later captured from cdc mode, and that could be problematic couldn't it be?

Yes, any deletes/updates past the position will be captured by CDC mode. 

> Where is the consistency lost? I feel like at worst we are doing some redundant updates / deletes / inserts. If the whole table is updated while snapshotting, we don't really loose information (as long as the cdc iterator finishes streaming all those updates).

The consistency is lost when a `DELETE` or `UPDATE` for a record which was never emitted as `SNAPSHOT` or `CREATE`. Lets consider this case:

Table `t` with initial values of 10<sup>min</sup> and 20<sup>max</sup>

1. Initial start, acquire values for snapshot $min=10$, $max=20$ and cdc pos<sub>m</sub>=20
2. Release lock 
3. Snapshot emits rows with IDs 10, 11, 12
4. In parallel Client deletes 14
7. Snapshot emits 13, 15, 16... 20. 
8. CDC starts, emits delete for key 14. 

Now the destination will need to coalesce a delete for a key which it never receives. 

The situation with `UPDATE` is somewhat more complicated because there won't be a record to update. Thus it is more of an upsert. 

Two things can help with this:
1. Release the global read lock when snapshot is complete.
2. Within the global read lock start a tx for each fetch worker with its own isolation snapshot. I believe the repeatable read / default should provide a consistent read snapshot. 

You are already doing 2) so we are good and can release the global lock earlier.

 --- 

Worst case scenario is when the snapshot is interrupted and needs to restart as records in the snapshot range can be affected and not seen by the snapshotter but seen by the cdc iterator, thus an update/delete will generate CDC event for a record which was never snapshotted. 

This is tangential as I'm not quite sure what kind of events MySQL emits, but in the case where a CDC event is a partial update (only returns values which were updated, not before/after), then data loss occurs, since only the values of the updated fields are provided:

1. R<sub>c</sub> (a,b,d): 1,2,3 
4. R<sub>u</sub> (a): 1 

In the above cases there is clear dependency between the events.

Reply:
Thanks for the detailed explanation @lyuboxa